src/formula/variant.cpp

```diff
@@ -647,121 +517,82 @@ variant variant::operator/(const variant& v) const
+		if(long_int%10 >= 5) {
 			long_int /= 10;
 			++long_int;
-		} else
+		} else {
 			long_int/=10;
+		}
 
-		return variant(  static_cast<int>(long_int) , variant::DECIMAL_VARIANT);
+		return variant(static_cast<int>(long_int), variant::DECIMAL_VARIANT);
 	}
 
-
 	const int numerator = as_int();
 	const int denominator = v.as_int();
+
 	if(denominator == 0) {
 		throw type_error("divide by zero error");
 	}
 
-	return variant(numerator/denominator);
+	return variant(numerator / denominator);
 }
 
 variant variant::operator%(const variant& v) const
 {
-	if(type_ == TYPE_DECIMAL || v.type_ == TYPE_DECIMAL) {
+	if(is_decimal() || v.is_decimal()) {
 		const int numerator = as_decimal();
 		const int denominator = v.as_decimal();
 		if(denominator == 0) {
 			throw type_error("divide by zero error");
 		}
 
-		return variant(numerator%denominator, DECIMAL_VARIANT);
+		return variant(numerator % denominator, DECIMAL_VARIANT);
 	} else {
 		const int numerator = as_int();
 		const int denominator = v.as_int();
 		if(denominator == 0) {
 			throw type_error("divide by zero error");
 		}
 
-		return variant(numerator%denominator);
+		return variant(numerator % denominator);
 	}
 }
 
 
 variant variant::operator^(const variant& v) const
 {
-	if( type_ == TYPE_DECIMAL || v.type_ == TYPE_DECIMAL ) {
+	if(is_decimal() || v.is_decimal()) {
 
-		double res = pow( as_decimal()/1000.0 , v.as_decimal()/1000.0 );
+		double res = pow(as_decimal()/1000.0 , v.as_decimal() / 1000.0);
 
 		if(res != res) return variant();
 
 		return variant(res, DECIMAL_VARIANT);
 	}
 
-	return variant(static_cast<int>(
-			round_portable(pow(static_cast<double>(as_int()), v.as_int()))));
+	return variant(static_cast<int>(round_portable(pow(static_cast<double>(as_int()), v.as_int()))));
 }
 
 variant variant::operator-() const
 {
-	if( type_ == TYPE_DECIMAL)
-		return variant( -decimal_value_, variant::DECIMAL_VARIANT );
+	if(is_decimal()) {
+		return variant(-as_decimal(), variant::DECIMAL_VARIANT);
+	}
 
 	return variant(-as_int());
 }
 
 bool variant::operator==(const variant& v) const
 {
-	if(type_ != v.type_) {
-		if( type_ == TYPE_DECIMAL || v.type_ == TYPE_DECIMAL ) {
+	if(type() != v.type()) {
+		if(is_decimal() || v.is_decimal()) {
 			return as_decimal() == v.as_decimal();
 		}
 
 		return false;
 	}
 
-	switch(type_) {
-	case TYPE_NULL: {
-		return v.is_null();
-	}
-
-	case TYPE_STRING: {
-		return *string_ == *v.string_;
-	}
-
-	case TYPE_INT: {
-		return int_value_ == v.int_value_;
-	}
-
-	case TYPE_DECIMAL: {
-		return decimal_value_ == v.decimal_value_;
-	}
-
-	case TYPE_LIST: {
-		if(num_elements() != v.num_elements()) {
-			return false;
-		}
-
-		for(size_t n = 0; n != num_elements(); ++n) {
-			if((*this)[n] != v[n]) {
-				return false;
-			}
-		}
-
-		return true;
-	}
-
-	case TYPE_MAP: {
-		return *map_ == *v.map_;
-	}
-
-	case TYPE_CALLABLE: {
-		return callable_->equals(v.callable_);
-	}
-	}
-
-	return false;
+	return value_ == v.value_;
```

Comment:
Um, hold on, this is wrong. With this, two strings will compare unequal even if the coincidentally have the same value, for example.

I'm guessing what you meant is something more like `return *value_ == *v.value_`? (Did `variant_value_base` implement these operators?)